packages/dev/core/test/unit/DeviceInput/babylon.deviceInput.test.ts

```diff
@@ -0,0 +1,226 @@
+        onInputChanged(DeviceType.Mouse, 0, mouseEvent);
+
+        expect(resultEvent).toEqual(mouseEvent);
+    });
+});
+
+describe("DeviceSourceManager", () => {
+    let engine: Nullable<NullEngine> = null;
+
+    beforeEach(function () {
+        engine = new NullEngine();
+        engine._deviceSourceManager = new MockInternalDeviceSourceManager() as InternalDeviceSourceManager;
+    });
+
+    afterEach(function () {
+        engine?.dispose();
+    });
+
+    it("should exist", () => {
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        expect(deviceSourceManager).not.toBe(null);
+        expect(engine?._deviceSourceManager?._refCount).toBe(1);
+    });
+
+    it("can use getDeviceSource", () => {
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const touchSource = new DeviceSource(deviceInputSystem, DeviceType.Touch, 0);
+        const touchSource2 = new DeviceSource(deviceInputSystem, DeviceType.Touch, 1);
+        const nullSource = deviceSourceManager.getDeviceSource(DeviceType.Touch, 0);
+        expect(nullSource).toBe(null);
+
+        deviceSourceManager._addDevice(touchSource);
+        deviceSourceManager._addDevice(touchSource2);
+
+        const firstAvailableSource = deviceSourceManager.getDeviceSource(DeviceType.Touch);
+        const specificSource = deviceSourceManager.getDeviceSource(DeviceType.Touch, 1);
+
+        expect(firstAvailableSource).toEqual(touchSource);
+        expect(specificSource).toEqual(touchSource2);
+
+        deviceSourceManager._removeDevice(DeviceType.Touch, 0);
+
+        const nextFirstSource = deviceSourceManager.getDeviceSource(DeviceType.Touch);
+        expect(nextFirstSource).toEqual(touchSource2);
+    });
+
+    it("can use getDeviceSources", () => {
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const touchSource = new DeviceSource(deviceInputSystem, DeviceType.Touch, 0);
+        const touchSource2 = new DeviceSource(deviceInputSystem, DeviceType.Touch, 1);
+        const emptyArray = deviceSourceManager.getDeviceSources(DeviceType.Touch);
+
+        expect(emptyArray.length).toBe(0);
+
+        deviceSourceManager._addDevice(touchSource);
+        deviceSourceManager._addDevice(touchSource2);
+
+        const touchArray = deviceSourceManager.getDeviceSources(DeviceType.Touch);
+        expect(touchArray.length).toBe(2);
+    });
+
+    it("can use onDeviceConnectedObservable", () => {
+        expect.assertions(1);
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const touchSource = new DeviceSource(deviceInputSystem, DeviceType.Touch, 0);

```

Comment:
This is a bit confusing. `new DeviceSourceManager` will create an `InternalDeviceSourceManager` which will create a real `DeviceInputSystem` right? Then, this is manually creating a `DeviceSource` with a mock `deviceInputSystem`? Am I reading this right?

Reply:
This section will be simplified and cleaned up